libs-haskell/bazel-runfiles/src/DA/Bazel/Runfiles.hs

```diff
@@ -27,27 +29,67 @@ exe :: FilePath -> FilePath
     | otherwise       = id
 
--- | Return the resources directory for the given runfiles dependency.
+data Resource = Resource
+  { runfilesPath :: FilePath
+  , resourcesPath :: FilePath
+  }
+  deriving (Eq, Ord, Show)
+
+-- | Return the path for the given resource dependency.
 --
--- In packaged application, using @bazel_tools/packaging/packaging.bzl@
--- this corresponds to the top-level @resources@ directory. In a @bazel run@ or
--- @bazel test@ target this corresponds to the @runfiles@ path of the given
--- @data@ dependency.
-locateRunfiles :: HasCallStack => FilePath -> IO FilePath
-locateRunfiles fp = do
+-- In a packaged application, using @bazel_tools/packaging/packaging.bzl@,
+--
+-- > locateResource Resource { runfilesPath, resourcesPath }
+--
+-- returns the path
+--
+-- > topLevelResourcesDir </> resourcesPath
+--
+-- where @topLevelResourcesDir@ is the top level @resources@ directory, a
+-- sibling of the packaged executable.
+--
+-- In a @bazel run@ or @bazel test@ target, the same expression returns
+--
+-- > topLevelResourceDir </> runfilesPath
+--
+-- where @topLevelResourceDir@ is the top level @runfiles@ directory assigned
+-- by Bazel.
+--
+locateResource :: HasCallStack => Resource -> IO FilePath
+locateResource Resource { runfilesPath, resourcesPath } = do
   -- If the current executable was packaged using @package_app@, then
   -- data files are stored underneath the resources directory.
   -- See @bazel_tools/packaging/packaging.bzl@.
-  -- This is based on Buck's runfiles behavior and users of locateRunfiles
-  -- expect the resources directory itself.
+  -- This is based on Buck's runfiles behavior.
   execPath <- getExecutablePath
   let jarResources = takeDirectory execPath </> "resources"
   hasJarResources <- doesDirectoryExist jarResources
   if hasJarResources
-      then pure jarResources
+      then pure (jarResources </> resourcesPath)
       else do
           runfiles <- Bazel.Runfiles.create
-          pure $! Bazel.Runfiles.rlocation runfiles fp
+          pure $! Bazel.Runfiles.rlocation runfiles runfilesPath
+
+-- | Return the resources directory for the given runfiles dependency.
+--
+-- In a packaged application, using @bazel_tools/packaging/packaging.bzl@
+-- this corresponds to the top-level @resources@ directory. In a @bazel run@ or
+-- @bazel test@ target this corresponds to the @runfiles@ path of the given
+-- @data@ dependency.
+--
+-- WARNING: In a packaged application this function __ignores the argument__
+-- and simply returns the resources directory itself - even if the
+-- corresponding @runfiles@ path is for a file.
+-- Prefer 'locateResource' for codepaths that will be used from both packaged
+-- applications and @bazel run@ or @bazel test@ targets.
+-- Users of this function should handle the result accordingly.
+--
+locateRunfiles :: HasCallStack => FilePath -> IO FilePath
+locateRunfiles runfilesPath =
+  locateResource Resource
+    { runfilesPath
+    , resourcesPath = ""
```

Comment:
Would defining `resourcesPath` as the `basename` of `runfilesPath` not work?

Reply:
not if we want to preserve the current behavior of `locateRunfiles` - just returning the base `"resources"` directory when running on a packaged app.

Reply:
I think we are talking slightly at crossed purposes here. I meant `basename` as in the `basename`/`dirname` shell commands , which work even when the path is a directly. For example:

```
~/daml$ pwd
/home/nic/daml
~/daml$ basename $(pwd)
daml
~/daml$ dirname $(pwd)
/home/nic
```

(And I'm pretty sure that Haskell `FilePath` has a similar concept.)